### DISCUSSION TITLE
`Type instantiation is excessively deep and possible infinite ts(2589)` after upgrading to `5.1` from `4.x`

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [mikro-orm/mikro-orm#2936] [Q&A] [answered]
**Describe the bug**
After upgrading to `5.1` there have been a handful of places where I've seen the `Type instantiation is excessively deep and possible infinite`. Most of the time it has to do with a helper type my team created to manage the constructor values of entities that extend a custom `BaseEntity`, but we just found our first one that seems to come from a pretty standard use of Mikro-orm.

I assume this likely has to do with our custom BaseEntity, but I'm not sure... seems like this would have come up more frequently with other users had it been using normal entities.

**Stack trace**
N/A

**To Reproduce**
``` typescript
const somethingId: any = "1";
const something = await em.findOne(Something, { id: somethingId });
```

On the snippet above, the `where` object has the red underline in VSCode and displays the error above.

**Expected behavior**
Type instantiation errors would not be thrown.

**Additional context**
We have a `Node` `BaseEntity` and we use the following helper methods to help define types for our constructor:
``` typescript
import { AnyNode, Node } from '../entities/Node';

export type SetOptional<Obj, Keys extends keyof Obj> = Omit<Obj, Keys> & {
  [K in Keys]?: Obj[K];
};

export type OmitMethods<Obj extends object> = {
  [K in keyof Obj as Obj[K] extends Function ? never : K]: Obj[K];
};

export type ConstructorValues<
  Entity extends AnyNode,
  RemovedKeys extends keyof OmitMethods<Omit<Entity, keyof Node<Entity>>> = never,
  OptionalKeys extends keyof OmitMethods<Omit<Entity, keyof Node<Entity> | RemovedKeys>> = never,
> = SetOptional<OmitMethods<Omit<Entity, keyof Node<Entity> | RemovedKeys>>, OptionalKeys>;
```

In `4.x` all of this worked flawlessly, but after upgrading, team members with less-performant machines are seeing the same error when using the `ConstructorValues` type (the right hand side is what triggers the error):
``` typescript
export type SomethingConstructorValues = ConstructorValues<Something, 'someOmittedField'>;
```

**Versions**

| Dependency | Version |
| - | - |
| node | `v16.13.0` |
| typescript | `^4.4.3` |
| mikro-orm | `5.1.1` |
| your-driver | `PostgreSQL` |

**Comment by B4nan** [comment]
I am moving this to discussions, as there is no reproduction. Without that I can't help much, I am not aware of anyone else having this problem. Will be happy to look into it if you can prepare something.

**Reply by SpencerKaiser** [reply]
Sounds good. What's the easiest way of creating a repro?

**Comment by SpencerKaiser** [comment, marked as answer]
@B4nan sorry for wasting your time... turns out everyone on my team except for me had the `JavaScript and TypeScript Nightly` extension installed in VSCode and were on `4.7.x` 🤦‍♂️ So yea... good lesson learned

Hopefully this will be helpful for others who might come across this while messing with Mikro-orm because this was the only package that triggered the issue. 

Thanks again and sorry for the wild goose chase!

**Reply by B4nan** [reply]
I can confirm that with 4.7 nightly I also see few errors like this in the codebase. Let's hope they will fix that before stable release.

**Reply by SpencerKaiser** [reply]
Yea I'd expect there would be with something so bleeding-edge 😉 

Also I know we were chatting on another issue about 5.1 and we finally ripped off the bandaid and made the switch... insanely big changes dude, you're crushing it 🚀
